File: dlna-dmr/src/main/java/com/android/cast/dlna/dmr/player/DLNARendererActivity.java
package com.android.cast.dlna.dmr.player;

import android.app.Service;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Bundle;
import android.os.IBinder;
import android.view.KeyEvent;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.Toast;
import android.widget.VideoView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.android.cast.dlna.dmr.DLNARendererService;
import com.android.cast.dlna.dmr.DLNARendererService.RendererServiceBinder;
import com.android.cast.dlna.dmr.R;

import org.fourthline.cling.model.types.UnsignedIntegerFourBytes;
import org.fourthline.cling.support.avtransport.lastchange.AVTransportVariable;
import org.fourthline.cling.support.model.Channel;
import org.fourthline.cling.support.model.TransportState;
import org.fourthline.cling.support.renderingcontrol.lastchange.ChannelVolume;
import org.fourthline.cling.support.renderingcontrol.lastchange.RenderingControlVariable;

/**
 *
 */
public class DLNARendererActivity extends AppCompatActivity {

    private static final String KEY_EXTRA_CURRENT_URI = "Renderer.KeyExtra.CurrentUri";

    public static void startActivity(Context context, String currentURI) {
        Intent intent = new Intent(context, DLNARendererActivity.class);
        intent.putExtra(KEY_EXTRA_CURRENT_URI, currentURI);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK); // start from service content,should add 'FLAG_ACTIVITY_NEW_TASK' flag.
        context.startActivity(intent);
    }

    private final UnsignedIntegerFourBytes INSTANCE_ID = new UnsignedIntegerFourBytes(0);
    private VideoView mVideoView;
    private ProgressBar mProgressBar;
    private CastMediaController mCastControlImp;
    private DLNARendererService mRendererService;

    private final ServiceConnection mServiceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            mRendererService = ((RendererServiceBinder) service).getRendererService();
            mRendererService.registerControlBridge(mCastControlImp = new CastMediaController(DLNARendererActivity.this, mRendererService));
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            mRendererService = null;
        }
    };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_dlna_renderer);
        bindService(new Intent(this, DLNARendererService.class), mServiceConnection, Service.BIND_AUTO_CREATE);
        mVideoView = findViewById(R.id.video_view);
        mProgressBar = findViewById(R.id.video_progress);
        openMedia(getIntent());
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        openMedia(intent);
    }

    private void openMedia(Intent intent) {
        Bundle bundle = intent.getExtras();
        if (bundle != null) {
            mProgressBar.setVisibility(View.VISIBLE);
            String currentUri = bundle.getString(KEY_EXTRA_CURRENT_URI);
            mVideoView.setVideoURI(Uri.parse(currentUri));
            mVideoView.setOnPreparedListener(new MediaPlayer.OnPreparedListener() {
                @Override
                public void onPrepared(MediaPlayer mp) {
                    mp.start();
                    mProgressBar.setVisibility(View.INVISIBLE);
                    notifyTransportStateChanged(TransportState.PLAYING);
                }
            });
            mVideoView.setOnErrorListener(new MediaPlayer.OnErrorListener() {
                @Override
                public boolean onError(MediaPlayer mp, int what, int extra) {
                    mProgressBar.setVisibility(View.INVISIBLE);
                    notifyTransportStateChanged(TransportState.STOPPED);
                    if (DLNARendererActivity.this != null) finish();
                    return true;
                }
            });
            mVideoView.setOnCompletionListener(new MediaPlayer.OnCompletionListener() {
                @Override
                public void onCompletion(MediaPlayer mp) {
                    mProgressBar.setVisibility(View.INVISIBLE);
                    notifyTransportStateChanged(TransportState.STOPPED);
                    if (DLNARendererActivity.this != null) finish();
                }
            });
        } else {
            Toast.makeText(this, "没有找到有效的视频地址，请检查...", Toast.LENGTH_SHORT).show();
            finish();
        }
    }

    @Override
    protected void onDestroy() {
        if (mVideoView != null) {
            mVideoView.stopPlayback();
        }
        notifyTransportStateChanged(TransportState.STOPPED);
        if (mRendererService != null) {
            mRendererService.unregisterControlBridge(mCastControlImp);
        }
        unbindService(mServiceConnection);
        super.onDestroy();
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        boolean handled = super.onKeyDown(keyCode, event);
        if (mRendererService != null) {
            if (keyCode == KeyEvent.KEYCODE_VOLUME_DOWN || keyCode == KeyEvent.KEYCODE_VOLUME_UP || keyCode == KeyEvent.KEYCODE_VOLUME_MUTE) {
                int volume = ((AudioManager) getApplication().getSystemService(Context.AUDIO_SERVICE)).getStreamVolume(AudioManager.STREAM_MUSIC);
                notifyRenderVolumeChanged(volume);
            } else if (keyCode == KeyEvent.KEYCODE_ENTER || keyCode == KeyEvent.KEYCODE_DPAD_CENTER) {
                if (mVideoView != null && mVideoView.isPlaying()) {
                    mVideoView.pause();
                    notifyTransportStateChanged(TransportState.PAUSED_PLAYBACK);
                } else if (mVideoView != null) {
                    mVideoView.resume();
                    notifyTransportStateChanged(TransportState.PLAYING);
                }
            }
        }
        return handled;
    }

    private void notifyTransportStateChanged(TransportState transportState) {
        if (mRendererService != null) {
            mRendererService.getAvTransportLastChange()
                    .setEventedValue(INSTANCE_ID, new AVTransportVariable.TransportState(transportState));
        }
    }

    private void notifyRenderVolumeChanged(int volume) {
        if (mRendererService != null) {
            mRendererService.getAudioControlLastChange()
                    .setEventedValue(INSTANCE_ID, new RenderingControlVariable.Volume(new ChannelVolume(Channel.Master, volume)));
        }
    }
}

File: dlna-dmr/src/main/java/com/android/cast/dlna/dmr/DLNARendererActivity.kt
package com.android.cast.dlna.dmr

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.media.AudioManager
import android.net.Uri
import android.os.Bundle
import android.os.IBinder
import android.view.KeyEvent
import android.view.View
import android.widget.ProgressBar
import android.widget.Toast
import android.widget.VideoView
import androidx.appcompat.app.AppCompatActivity
import com.android.cast.dlna.dmr.R.id
import com.android.cast.dlna.dmr.R.layout
import com.android.cast.dlna.dmr.RenderControl.VideoViewRenderControl
import org.fourthline.cling.model.types.UnsignedIntegerFourBytes
import org.fourthline.cling.support.avtransport.lastchange.AVTransportVariable
import org.fourthline.cling.support.model.Channel.Master
import org.fourthline.cling.support.model.TransportState
import org.fourthline.cling.support.model.TransportState.*
import org.fourthline.cling.support.renderingcontrol.lastchange.ChannelVolume
import org.fourthline.cling.support.renderingcontrol.lastchange.RenderingControlVariable.Volume

/**
 *
 */
class DLNARendererActivity : AppCompatActivity() {

    companion object {
        private const val KEY_EXTRA_CURRENT_URI = "Renderer.KeyExtra.CurrentUri"
        fun startActivity(context: Context, currentURI: String?) {
            val intent = Intent(context, DLNARendererActivity::class.java)
            intent.putExtra(KEY_EXTRA_CURRENT_URI, currentURI)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK) // start from service content,should add 'FLAG_ACTIVITY_NEW_TASK' flag.
            context.startActivity(intent)
        }
    }

    private val instanceId = UnsignedIntegerFourBytes(0)
    private val videoView: VideoView? by lazy { findViewById(id.video_view) }
    private val progressBar: ProgressBar? by lazy { findViewById(id.video_progress) }
    private var rendererService: DLNARendererService? = null

    private val serviceConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            //TODO: can not access to RendererServiceBinder.
            //rendererService = (service as RendererServiceBinder).rendererService
            rendererService?.setRenderControl(VideoViewRenderControl(videoView!!))
        }

        override fun onServiceDisconnected(name: ComponentName) {
            rendererService = null
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layout.activity_dlna_renderer)
        bindService(Intent(this, DLNARendererService::class.java), serviceConnection, BIND_AUTO_CREATE)
        openMedia(intent)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        openMedia(intent)
    }

    private fun openMedia(intent: Intent) {
        val bundle = intent.extras
        if (bundle != null) {
            progressBar?.visibility = View.VISIBLE
            val currentUri = bundle.getString(KEY_EXTRA_CURRENT_URI)
            videoView?.setVideoURI(Uri.parse(currentUri))
            videoView?.setOnPreparedListener { mp ->
                mp.start()
                progressBar?.visibility = View.INVISIBLE
                notifyTransportStateChanged(PLAYING)
            }
            videoView!!.setOnErrorListener { mp, what, extra ->
                progressBar?.visibility = View.INVISIBLE
                notifyTransportStateChanged(STOPPED)
                finish()
                true
            }
            videoView!!.setOnCompletionListener {
                progressBar?.visibility = View.INVISIBLE
                notifyTransportStateChanged(STOPPED)
                finish()
            }
        } else {
            Toast.makeText(this, "没有找到有效的视频地址，请检查...", Toast.LENGTH_SHORT).show()
            finish()
        }
    }

    override fun onDestroy() {
        videoView?.stopPlayback()
        notifyTransportStateChanged(STOPPED)
        unbindService(serviceConnection)
        super.onDestroy()
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        val handled = super.onKeyDown(keyCode, event)
        if (rendererService != null) {
            if (keyCode == KeyEvent.KEYCODE_VOLUME_DOWN || keyCode == KeyEvent.KEYCODE_VOLUME_UP || keyCode == KeyEvent.KEYCODE_VOLUME_MUTE) {
                val volume = (application.getSystemService(AUDIO_SERVICE) as AudioManager).getStreamVolume(AudioManager.STREAM_MUSIC)
                notifyRenderVolumeChanged(volume)
            } else if (keyCode == KeyEvent.KEYCODE_ENTER || keyCode == KeyEvent.KEYCODE_DPAD_CENTER) {
                if (videoView?.isPlaying == true) {
                    videoView?.pause()
                    notifyTransportStateChanged(PAUSED_PLAYBACK)
                } else if (videoView != null) {
                    videoView?.resume()
                    notifyTransportStateChanged(PLAYING)
                }
            }
        }
        return handled
    }

    private fun notifyTransportStateChanged(transportState: TransportState) {
        rendererService?.avTransportLastChange?.setEventedValue(instanceId, AVTransportVariable.TransportState(transportState))
    }

    private fun notifyRenderVolumeChanged(volume: Int) {
        rendererService?.audioControlLastChange?.setEventedValue(instanceId, Volume(ChannelVolume(Master, volume)))
    }

}